clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java

```diff
@@ -47,32 +190,166 @@ public boolean isValid() {
-     * will block for at most maxWaitTimeMs. It is expected that this method be called from application thread
-     * contexts (ie. through Producer.send). The PID it self will be retrieved in the background thread.
-     * @param maxWaitTimeMs The maximum time to block.
-     * @return a PidAndEpoch object. Callers must call the 'isValid' method fo the returned object to ensure that a
-     *         valid Pid and epoch is actually returned.
-     */
-    public synchronized PidAndEpoch awaitPidAndEpoch(long maxWaitTimeMs) throws InterruptedException {
-        long start = time.milliseconds();
-        long elapsed = 0;
-        while (!hasPid() && elapsed < maxWaitTimeMs) {
-            wait(maxWaitTimeMs);
-            elapsed = time.milliseconds() - start;
+    public boolean isFenced() {
+        return isFenced;
+    }
+
+    public void beginTransaction() {
+        isInTransaction = true;
+    }
+
+    public boolean isCompletingTransaction() {
+        return isInTransaction && isCompletingTransaction;
+    }
+
+    public synchronized FutureTransactionalResult beginCommittingTransaction() {
+        return beginCompletingTransaction(true);
+    }
+
+    public synchronized FutureTransactionalResult beginAbortingTransaction() {
+        return beginCompletingTransaction(false);
+    }
+
+    private FutureTransactionalResult beginCompletingTransaction(boolean isCommit) {
+        if (!isCompletingTransaction) {
+            TransactionalRequestResult result = new TransactionalRequestResult();
+            FutureTransactionalResult resultFuture = new FutureTransactionalResult(result);
+
+            isCompletingTransaction = true;
+            if (!newPartitionsToBeAddedToTransaction.isEmpty()) {
+                pendingTransactionalRequests.add(addPartitionsToTransactionRequest(false));
+            }
+            pendingTransactionalRequests.add(endTxnRequest(isCommit, false, result));
+            return resultFuture;
         }
-        return pidAndEpoch;
+        return null;
+    }
+
+    public synchronized FutureTransactionalResult sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets,
+                                         String consumerGroupId) {
+        TransactionalRequestResult result = new TransactionalRequestResult();
+        FutureTransactionalResult resultFuture = new FutureTransactionalResult(result);
+        pendingTransactionalRequests.add(addOffsetsToTxnRequest(offsets, consumerGroupId, false, result));
+        return resultFuture;
+    }
+
+    public boolean isInTransaction() {
+        return isTransactional() && isInTransaction;
+    }
+
+
+    public synchronized void maybeAddPartitionToTransaction(TopicPartition topicPartition) {
+        if (partitionsInTransaction.contains(topicPartition))
+            return;
+        newPartitionsToBeAddedToTransaction.add(topicPartition);
+    }
+
+    public void needsRetry(TransactionalRequest request) {
```

Comment:
As with the others, these are only called in the context of the background thread (in particular from `Sender.maybeSendTransactionalRequest()`), hence no need for synchronization.